File: FSharpTracer/FSharpTracer/Surface.fs
namespace FSharpTracer

type Surface = 
   abstract Roughness : unit -> float
   abstract Diffuse : Vector -> Color
   abstract Specular : Vector -> Color
   abstract Reflect : Vector -> float

type Shiny() = 
   interface Surface with
      member s.Roughness() = 50.0
      
      member s.Diffuse _ = 
         { R = 1.0
           G = 1.0
           B = 1.0 }
      
      member s.Specular _ = 
         { R = 0.5
           G = 0.5
           B = 0.5 }
      
      member s.Reflect _ = 0.3

type Checkerboard() = 
   interface Surface with
      member s.Roughness() = 150.0
      
      member s.Diffuse v = 
         if (floor v.X + floor v.Z) % 2.0 <> 0.0 then
            { R = 1.0
              G = 1.0
              B = 1.0 }
         else
            { R = 0.0
              G = 0.0
              B = 0.0 }
      
      member s.Specular _ = 
         { R = 1.0
           G = 1.0
           B = 1.0 }
      
      member s.Reflect v = 
         if (floor v.X + floor v.Z) % 2.0 <> 0.0 then
            1.0
         else
            0.7
      

File: SwiftTracer/SwiftTracer/Surface.swift
import Foundation

protocol Surface {
    func Roughness() -> Double
    func Diffuse(Vector) -> Color
    func Specular(Vector) -> Color
    func Reflect(Vector) -> Double
}

class Shiny : Surface {
    func Roughness() -> Double {
        return 50.0
    }
    
    func Diffuse(_: Vector) -> Color {
        return Color(r:1.0, g:1.0,b:1.0)
    }
    
    func Specular(_: Vector) -> Color {
        return Color(r:0.5,g:0.5,b:0.5)
    }
    func Reflect(_: Vector) -> Double {
        return 0.3
    }
}

class Checkerboard : Surface {
    func Roughness() -> Double {
        return 150.0
    }
    
    func Diffuse(v: Vector) -> Color {
        if (floor(v.x) + floor(v.z)) % 2 != 0.0 {
            return Color(r:1.0, g:1.0,b:1.0)
        } else {
            return Color(r:0.0, g:0.0,b:0.0)
        }
    }
    
    func Specular(_: Vector) -> Color {
        return Color(r:1.0, g:1.0,b:1.0)
    }
    func Reflect(v: Vector) -> Double {
        if (floor(v.x) + floor(v.z)) % 2 != 0.0 {
            return 1.0
        } else {
            return 0.7
        }
    }
}